### ISSUE TITLE
[$250] Xero - Infinite spinner on "Waiting for imported data to load" when connecting to Xero

### ISSUE THREAD

**Issue by trjExpensify** [Expensify/App#53674]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v9.0.72-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** N/A 
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):** tom+xerodofl@trj.chat
**Logs:** [here](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%22KkYmhu%22)+AND+timestamp:[2024-12-05T23:50:16.000Z+TO+2024-12-06T01:50:21.000Z]&sort=asc&size=2000) (internal). 

```
Couldn't get valid connection data ~~ integration: 'xero' policyID: '75A37FFA15152001' data: 'is_temp_logs2fc8f4db-2e06-4013-adab-8c26a017556f.log' class: 'ExpensifyException' msg: 'Invalid connection configuration' code: '410'
```
```
Couldn't load Xero connection data ~~ policyID: '75A37FFA15152001' class: 'ExpensifyException' msg: 'Invalid connection configuration' code: '410'
```

**Expensify/Expensify Issue URL:**
**Issue reported by:** @trjExpensify 
**Slack conversation** (hyperlinked to channel name): [#convert thread](https://expensify.slack.com/archives/C07HPDRELLD/p1733447022894439). 

## Action Performed:
1. Sign-up for a new account on new.expensify.com 
2. Choose "Manage my team's expenses" > 1-10 > Xero 
3. Connect the workspace to Xero (validate the account and add a 2FA code)
4. Observe the infinite loading spinner for "Waiting for imported data to load"
5. Navigate away from the Accounting page and back again 
6. Observe the connection has loaded. 

## Expected Result:
You shouldn't have to navigate away and back for the connection details to appear. 

## Actual Result:
Infinite loading spinner.  

## Workaround:
Yes, navigate away and back again. 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/eab931d2-0759-4385-98aa-c9728e7fc989

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021864838632082983296</li>
        <li>Upwork Job ID: 1864838632082983296</li>
        <li>Last Price Increase: 2024-12-13</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @lakchote</details>

**Comment by trjExpensify** [comment]
CC: @hungvu193 @lakchote as you both have experience in this domain!

**Comment by hungvu193** [comment]
I can reproduce. It looks like `policy?.connections` is empty so the loading indicator is still visible, possibly BE issue, our Pusher event didn't send the connections data:

<img width="747" alt="Screenshot 2024-12-06 at 10 15 04" src="https://github.com/user-attachments/assets/45472c87-b8fa-41bc-92b8-26d9d25b8d34">

**Comment by trjExpensify** [comment]
Interesting! @lakchote would this perhaps have something to do with turning off OnyxUpdates for Classic? I didn't think it would, because `shouldHandleOnyxUpdates` is set to `true` for job source NewDot. 

https://github.com/Expensify/Integration-Server/pull/7934/files
https://github.com/Expensify/Web-Expensify/pull/42292/files

**Comment by lakchote** [comment]
I've followed the exact testing steps outlined in the issue and couldn't reproduce.

The error `Invalid connection configuration` comes when there are required properties missing from the Xero's configuration keys such as `mappings`, `importTrackingCategories`, `export`...

@francoisl do you have any hints on when this edge case could happen?

**Comment by trjExpensify** [comment]
>I've followed the exact testing steps outlined in the issue and couldn't reproduce.

Interesting that both Hans and I could repro. 🤔 I can try again.

**Comment by trjExpensify** [comment]
Yep, same again. I'm using the 1Password credentials for `SRAM Test Xero`. 

https://github.com/user-attachments/assets/f9f99ec3-3eb4-4efe-bfe3-3ff8caec00f9

**Comment by trjExpensify** [comment]
>I've followed the exact testing steps outlined in the issue and couldn't reproduce.

@lakchote was this staging or dev only?

**Comment by trjExpensify** [comment]
>I've just tried on staging and I now have a Bad gateway error when trying to connect, do you have one too?

I didn't get this one. Just an infinite loader on "Waiting for imported data to load"...

**Comment by lakchote** [comment]
There has been no new changes related to Xero's code nor Integration-Server recently, I'll need to investigate it further.

**Comment by trjExpensify** [comment]
@lakchote is going to investigate this a bit, Melv.

**Comment by lakchote** [comment]
I've retested the whole flow on another account on staging and the issue remains.

Same on dev, I actually get the same issue now.

The Onyx update for the policy is never sent until we go back to reenter which calls `OpenPolicyProfilePage`:
<img width="432" alt="image" src="https://github.com/user-attachments/assets/ef61aad5-d776-4cfd-bb99-e5bb809ab5f9">

@francoisl are you aware of any changes related to Onyx/Pusher in Integration-Server?

**Comment by francoisl** [comment]
I tested and was able to repro on dev too. I'm not aware of any changes in Integration-Server, though in this case I imagine the Onyx updates should come from Auth when we save the connection details in the policy?

**Comment by flodnv** [comment]
> Nvm, it doesn't seem to be the root cause of the issue.

Agreed, `log` means that we do this: https://github.com/Expensify/Auth/blob/4345e99bff32f89093e2ad742274f0d4ee62e883/auth/command/SavePolicy.cpp#L483

**Comment by lakchote** [comment]
Asked for external eyes on this here: https://expensify.slack.com/archives/C03TQ48KC/p1734108170237789

**Comment by trjExpensify** [comment]
@lakchote should we assign you here given that you're investigating it?

**Comment by lakchote** [comment]
> @lakchote should we assign you here given that you're investigating it?

Yes, I've assigned myself.

**Comment by lakchote** [comment]
PR is under review.

**Comment by trjExpensify** [comment]
Great stuff!

**Comment by trjExpensify** [comment]
Bit of a slowdown over the holidays. Review comment for @lakchote to take a look at [here](https://github.com/Expensify/Auth/pull/13570#issuecomment-2560386874).

**Comment by lakchote** [comment]
> Bit of a slowdown over the holidays. Review comment for @lakchote to take a look at [here](https://github.com/Expensify/Auth/pull/13570#issuecomment-2560386874).
>

The reason we didn't get the sync updates was because of [this](https://expensify.slack.com/archives/C03SC9DVD/p1734730555324959).

It's now fixed (I've tried on staging).

I'll let you test on your side @trjExpensify before closing this issue.

**Comment by trjExpensify** [comment]
Commented in the PR. We were supposed to be protected against whenever we turn off onyx updates for OldDot to fight bugs and performance per here: https://github.com/Expensify/App/issues/53674#issuecomment-2522895430

So I think something broke that meant that NewDot users connecting accounting solutions were impacted this time when we did it.

**Comment by lakchote** [comment]
> Commented in the PR. We were supposed to be protected against whenever we turn off onyx updates for OldDot to fight bugs and performance per here: [#53674 (comment)](https://github.com/Expensify/App/issues/53674#issuecomment-2522895430)
> 
> So I think something broke that meant that NewDot users connecting accounting solutions were impacted this time when we did it.
>

These two, `shouldHandleOnyxUpdates` being set to `true` for NewDot and the `SendFetchableOnyxUpdates` job being disabled are different things.

`shouldHandleOnyxUpdates` being set to `true` will allow adding Onyx updates to the response (https://github.com/Expensify/Auth/blob/59aaeec11e444198a61d605cec5786acc2f9715f/auth/AuthCommand.cpp#L845-L847).

However in the case of fetchable Onyx updates, we still need the job to be enabled in order for it to be created and the Onyx updates to be sent. 

Why?
Because of the flow for integrations:
**Web** -> `ConnectPolicyToXero` makes a call to the IS to connect the policy
**IS** -> do processing -> calls at the end `Policy_Connection_Set` (**OldDot**) with param `shouldHandleOnyxUpdates` set to `true`
**Web** -> `Policy_Connection_Set` (**OldDot**) calls Auth command `SavePolicy`

It's in OldDot that we trigger the job creation (see section `Processing the event data returned from Auth in Web` for more details).
 
It was manually disabled and as such, we couldn't send the Onyx updates that were being added.

And they were added, see in the logs reported in the GH's issue body:
![image](https://github.com/user-attachments/assets/28529240-914b-41f6-a158-cd52b8ac3bf2)

**Comment by trjExpensify** [comment]
>These two, shouldHandleOnyxUpdates being set to true for NewDot and the SendFetchableOnyxUpdates job being disabled are different things.

Interesting. So before when we protected accounting connections originating from NewDot, you suspect we didn't disable `SendFetchableOnyxUpdates` that time, or that didn't exist at the time or something? 

I think we need to think about how we can keep `SendFetchableOnyxUpdates` enabled for NewDot requests, in the same way we did for `shouldHandleOnyxUpdates` when the job source is NewDot [here](https://github.com/Expensify/Integration-Server/blob/204fba49ffa07730bbcfa9e84bc0d5fb4cb23708/src/expensify/apiCommands/ExpensifyAPICommand.java#L615-L619). 

CC: @aldo-expensify for vis on this discussion as well btw

**Comment by aldo-expensify** [comment]
> So before when we protected accounting connections originating from NewDot, you suspect we didn't disable `SendFetchableOnyxUpdates` that time, or that didn't exist at the time or something?

From the conversation, I'm also understanding that.

> I think we need to think about how we can keep `SendFetchableOnyxUpdates` enabled for NewDot requests, in the same way we did for `shouldHandleOnyxUpdates` when the job source is NewDot [here](https://github.com/Expensify/Integration-Server/blob/204fba49ffa07730bbcfa9e84bc0d5fb4cb23708/src/expensify/apiCommands/ExpensifyAPICommand.java#L615-L619).

Agree

**Comment by lakchote** [comment]
> > These two, shouldHandleOnyxUpdates being set to true for NewDot and the SendFetchableOnyxUpdates job being disabled are different things.
> 
> Interesting. So before when we protected accounting connections originating from NewDot, you suspect we didn't disable `SendFetchableOnyxUpdates` that time, or that didn't exist at the time or something?
>

I'm not sure of clearly understanding this sentence, could you please rephrase it?
As long as there were fetchable Onyx updates, `SendFetchableOnyxUpdates` job must have been enabled to receive Onyx updates in NewDot for the integrations, otherwise we would have never received them.

To confirm, I'm cc'ing one of the best experts related to this area of code, @iwiznia could we have your input on this?
(see https://github.com/Expensify/App/issues/53674#issuecomment-2572765665 for a quick recap of what we're discussing).

**Comment by aldo-expensify** [comment]
hmm when we disabled the onyx updates coming from other sources (other than NewDot) here: https://github.com/Expensify/Web-Expensify/pull/44643/files, I think we should have done:

```diff
+        $headers['shouldHandleOnyxUpdates'] = true;
-        $headers['shouldHandleOnyxUpdates'] = ($_REQUEST['shouldHandleOnyxUpdates'] ?? false) === 'true' || CI_RUNNING || Request::isECashApp() || str_starts_with($_REQUEST['command'] ?? '', 'ChatBot_');
```

instead of

```diff
+        $headers['shouldHandleOnyxUpdates'] = true;
-        $headers['shouldHandleOnyxUpdates'] = Request::isECashApp() || PHP_UNIT;
```

The `$_REQUEST['shouldHandleOnyxUpdates'] ?? false) === 'true'` is needed (and was added in [this PR](https://github.com/Expensify/Web-Expensify/pull/42292/files#diff-7e2d65062b3d1b62158a39b82faabb6666d1fc9a64abe64404c94c374208fb56L460)) to respect the  `shouldHandleOnyxUpdates` set in the IS:

https://github.com/Expensify/Integration-Server/blob/204fba49ffa07730bbcfa9e84bc0d5fb4cb23708/src/expensify/apiCommands/ExpensifyAPICommand.java#L615-L619


The long line `($_REQUEST['shouldHandleOnyxUpdates'] ?? false) === 'true' || CI_RUNNING || Request::isECashApp() || str_starts_with($_REQUEST['command'] ?? '', 'ChatBot_');` was used when we disabled onyx updates the previous time.

**Comment by iwiznia** [comment]
> As long as there were fetchable Onyx updates, SendFetchableOnyxUpdates job must have been enabled to receive Onyx updates in NewDot for the integrations, otherwise we would have never received them.

Correct, if SendFetchableOnyxUpdates was disabled for non newDot requests, then any updates inserted from non newDot would not be delivered live (they would be downloaded by the app when it detects a gap)

**Comment by lakchote** [comment]
> The $_REQUEST['shouldHandleOnyxUpdates'] ?? false) === 'true' is needed (and was added in [this PR](https://github.com/Expensify/Web-Expensify/pull/42292/files#diff-7e2d65062b3d1b62158a39b82faabb6666d1fc9a64abe64404c94c374208fb56L460)) to respect the shouldHandleOnyxUpdates set in the IS:
>

Right now it's set to `true` by default, but I do agree that we should put back the logic you've mentioned to take into account the value sent from IS.

> Correct, if SendFetchableOnyxUpdates was disabled for non newDot requests, then any updates inserted from non newDot would not be delivered live (they would be downloaded by the app when it detects a gap)

Thanks for your answer.

So, it makes sense why we didn't get Onyx updates since it is an OldDot request at the time where we set the policy's connection data (please refer to the flow [here](https://github.com/Expensify/App/issues/53674#issuecomment-2572765665)). I'll be also detailing more below.

In regards to what you previously said @trjExpensify:
> I think we need to think about how we can keep SendFetchableOnyxUpdates enabled for NewDot request
>
The request to connect Xero (or any other integration) from NewDot, makes a call to the IS which in turns makes a call to OldDot to set the policy connection data. 

As a result, the referer set is `integration` and not `ecash` (the equivalent value for a NewDot request):
![image](https://github.com/user-attachments/assets/3d9f54e5-45a9-4b35-b6e3-abfecc791c17)

As such, the request will not be considered a NewDot request and, when `SavePolicy` is called, the updates inserted will not be delivered live if the `SendFetchableOnyxUpdates` job is disabled for OldDot requests.

**Comment by lakchote** [comment]
@iwiznia can we _always_ create the `SendFetchableOnyxUpdates` job even if we've disabled it for OldDot for the specific case where the request originates from IS and the source is NewDot? We discussed this with @trjExpensify, the reason being it would prevent having this important side effect (no indication of sync progress) for NewDot accounting integrations, when the job `SendFetchableOnyxUpdates` is disabled for OldDot.

**Comment by iwiznia** [comment]
Maybe we can, but we also are not planning on disabling it again for now.

**Comment by trjExpensify** [comment]
I don't think we planned to disable it again the last time either really. 😅

**Comment by lakchote** [comment]
Well, I've put up a [PR](https://github.com/Expensify/Web-Expensify/pull/45021) to do that and tagged both @iwiznia and @aldo-expensify on it, in case we all agree upon going into that direction.

**Comment by trjExpensify** [comment]
Okay great, so the outcome here is basically accounting for this in the code we write if we need to disable `SendFetchableOnyxUpdates` again in the future. Will we remember that, or is there somewhere helpful to write this down?

**Comment by lakchote** [comment]
> ```
> if (
>     Request::isECashApp() ||
>     PHP_UNIT ||
>     // Allow integration updates from NewDot to pass
>     (Request::isIntegrationServer() && ($_REQUEST['shouldHandleOnyxUpdates'] ?? false) === 'true')
> ) {
>      // ...
>     Onyx::scheduleSendFetchableJob($pusherSocketID, $json['fetchableOnyxUpdates'] ?? []
>     );
> }
> ```

I've created a [SO](https://stackoverflowteams.com/c/expensify/questions/20817/20818#20818) post, to make sure we'll remember it.

**Comment by trjExpensify** [comment]
Thanks, so I think we're all set here. Closing!
